man/nxsfromxml.1
.TH nxsfromxml 1 "2014-01-02" nxsfromxml
.SH NAME
.B nxsfromxml
\- command-line tool for creating NeXus files from XML template

.SH SYNOPSIS
.B  nxsfromxml
.I [-h] [-x XMLFILE] [-p PARENT] [-d DATA] [-j JSONFILE] [-t STIME] [-n NRECORDS] [-v] [-a] [xml_config]


.SH DESCRIPTION
.B nxsfromxml
is a command-line tool which uses the nxswriter package to create NeXus files from XML template. Optional data can be provided in a json file or a json string  as a command-line parameter.

.SH OPTIONS
.IP "xml_config"
nexus writer configuration string

optional arguments:
.IP "-h, --help"
show this help message and exit
.IP "-x XMLFILE, --xml-file XMLFILE"
optional file with nexus configuration string
.IP "-p PARENT, --parent PARENT"
nexus file name and optional parent nexus path
.IP "-d DATA, --data DATA"
json string with data
.IP "-j JSONFILE, --json-file JSONFILE"
json data file in a jsonrecord attribute format
.IP "-t STIME, --time STIME"
time between record steps in seconds, default: 1 s
.IP "-n NRECORDS, --nrecords NRECORDS"
number of performed record steps
.IP "-v, --verbose"
verbose mode
.IP "-a, --append"
append mode

.SH EXAMPLES:
  nxsfromxml  -x config.xml -p newfile.nxs -j datastring.json

    - creates a newfile.nxs file with configuration from config.xml and data from datastring.json

  nxsfromxml  -x config.xml -p outputfile.nxs -d '{"sample_name":"sn", "chemical_formula":"H2O", "beamtime_id":"1234", "title":"Sample"}'

    - creates an outputfile.nxs file with configuration from config.xml and data from given json dictionary

  nxsconfig get source mono slit1 mirror2 | nxsfromxml -p basicsetup.nxs

    - creates an basicsetup.nxs file with configuration of source mono slit1 mirror2 components created by nxsconfig get 

.SH SEE ALSO
https://github.com/jkotan/nexdatas/

.SH COPYRIGHT
Copyrights (c) 2020, GNU GPL v2, DESY, Jan Kotanski

.SH BUGS
Please report bugs on the project's mailing list:
mailto://jankotan@gmail.com

.SH AUTHORS
Jan Kotanski <jankotan@gmail.com>



ChangeLog
2020-05-08 Jan Kotanski <jankotan@gmail.com>
	* fix typo in printout
	* tagged as 2.28.2

2020-05-08 Jan Kotanski <jankotan@gmail.com>
	* fix manifest file
	* tagged as 2.28.1

2020-05-08 Jan Kotanski <jankotan@gmail.com>
	* add nxsfromxml script
	* add optional parent nexus path to fileName attribute
	* tagged as 2.28.0

2020-04-14 Jan Kotanski <jankotan@gmail.com>
	* speed up getJSONData
	* tagged as 2.27.0

2020-03-06 Jan Kotanski <jankotan@gmail.com>
	* support lower record names for client datasources
	* tagged as 2.26.0

2020-01-15 Jan Kotanski <jankotan@gmail.com>
	* fix reshaping LIMA encoded attributes
	* tagged as 2.25.3

2020-01-07 Jan Kotanski <jankotan@gmail.com>
	* fix swapping endianness for LIMA encoded attributes
	* tagged as 2.25.2

2019-10-10 Jan Kotanski <jankotan@gmail.com>
	* add pip support
	* change to setuptools
	* tagged as 2.25.1

2019-10-01 Jan Kotanski <jankotan@gmail.com>
	* add AddingLogs property
	* tagged as 2.25.0

2019-09-24 Jan Kotanski <jankotan@gmail.com>
	* add external filters
	* fixes for no-tango mode
	* tagged as 2.24.1

2019-09-05 Jan Kotanski <jankotan@gmail.com>
	* add external filters
	* requires nxstools 2.67.0
	* tagged as 2.24.0

2019-07-11 Jan Kotanski <jankotan@gmail.com>
	* remove strip attribute
	* tagged as 2.23.1

2019-06-27 Jan Kotanski <jankotan@gmail.com>
	* change minidom to etree
	* remove circular dependences by adding weakref's
	* tagged as 2.23.0

2019-03-18 Jan Kotanski <jankotan@gmail.com>
	* workaround for the case when Client data is None
	* tagged as 2.22.6

2019-02-28 Jan Kotanski <jankotan@gmail.com>
	* fix for links to not existing object
	* tagged as 2.22.5

2019-02-28 Jan Kotanski <jankotan@gmail.com>
	* allows creating links to not existing object
	* tagged as 2.22.4

2019-01-15 Jan Kotanski <jankotan@gmail.com>
	* add flush for stderr stream
	* tagged as 2.22.3

2018-11-26 Jan Kotanski <jankotan@gmail.com>
	* workaround for broken pipe in StreamSet
	* tagged as 2.22.2

2018-11-21 Jan Kotanski <jankotan@gmail.com>
	* fix tests for buster
	* fix for PYTG_BUG_213
	* tagged as 2.22.1

2018-10-18 Jan Kotanski <jankotan@gmail.com>
	* split the package
	* fix for h5cppwriter module
	* tagged as 2.22.0

2018-10-09 Jan Kotanski <jankotan@gmail.com>
	* fixes for SWMR with h5cpp
	* tagged as 2.21.1

2018-09-20 Jan Kotanski <jankotan@gmail.com>
	* add support for python3
	* tagged as 2.21.0

2018-09-04 Jan Kotanski <jankotan@gmail.com>
	* add support for fqdn in TangoServer
	* tagged as 2.20.0

2018-08-15 Jan Kotanski <jankotan@gmail.com>
	* add support for h5cpp
	* tagged as 2.19.0

2018-05-18 Jan Kotanski <jankotan@gmail.com>
	* second step in  support for utf8 in h5pywriter
	* tagged as 2.18.2

2018-05-18 Jan Kotanski <jankotan@gmail.com>
	* fix support for utf8 in h5pywriter
	* tagged as 2.18.1

2018-02-01 Jan Kotanski <jankotan@gmail.com>
	* set libver=latest if swmr is True
	* add reopen, names, for FileWriter objects
	* add get_links for FileWriter
	* fix refresh method for FileWriter objects
	* tagged as 2.18.0

2017-11-21 Jan Kotanski <jankotan@gmail.com>
	* set DefaultCanFail property to True
	* add error on shape mismatch
	* tagged as 2.17.1

2017-11-20 Jan Kotanski <jankotan@gmail.com>
	* add DefaultCanFail property
	* tagged as 2.17.0

2017-09-26 Jan Kotanski <jankotan@gmail.com>
	* do not create a link when target is empty
	* tagged as 2.16.1

2017-09-08 Jan Kotanski <jankotan@gmail.com>
	* append errors in markFail
	* tagged as 2.16.0

2017-07-12 Jan Kotanski <jankotan@gmail.com>
	* add writer parameter into open_file/create_file
	* allow to run two different writers simultaneously
	* tagged as 2.15.0

2017-07-12 Jan Kotanski <jankotan@gmail.com>
	* improve thread safety of NXSDataWriter
	* tagged as 2.14.3

2017-06-28 Jan Kotanski <jankotan@gmail.com>
	* comment printouts
	* tagged as 2.14.2

2017-06-23 Jan Kotanski <jankotan@gmail.com>
	* move globallock into __init__
	* tagged as 2.14.1

2017-06-23 Jan Kotanski <jankotan@gmail.com>
	* add skipacqusiotion attribute
	* add globallock to nxswriter
	* tagged as 2.14.0

2017-06-20 Jan Kotanski <jankotan@gmail.com>
	* change Streams module into StreamSet class
	* tagged as 2.13.0

2017-06-13 Jan Kotanski <jankotan@gmail.com>
	* add global canfail attribute
	* tagged as 2.12.0

2017-05-24 Jan Kotanski <jankotan@gmail.com>
	* fix for reopening logs
	* tagged as 2.11.1

2017-05-11 Jan Kotanski <jankotan@gmail.com>
	* add file split with stepsPerFile
	* tagged as 2.11.0

2017-04-22 Jan Kotanski <jankotan@gmail.com>
	* avoid closing file when state is ON or OFF
	* tagged as 2.10.2

2017-04-13 Jan Kotanski <jankotan@gmail.com>
	* fix for file cannot be closed
	* tagged as 2.10.1

2017-04-07 Jan Kotanski <jankotan@gmail.com>
	* add STEP block recording
	* nexus_configuration_logs to nexus_logs and configuration
	* tagged as 2.10.0

2017-03-22 Jan Kotanski <jankotan@gmail.com>
	* change Nexus__entry__%s_XML to nexus__entry__%s_xml
	* tagged as 2.9.0

2017-03-22 Jan Kotanski <jankotan@gmail.com>
	* speed up type convetion
	* tagged as 2.8.1

2017-03-21 Jan Kotanski <jankotan@gmail.com>
	* fix type convetion
	* tagged as 2.8.0

2017-03-19 Jan Kotanski <jankotan@gmail.com>
	* fix manpage
	* tagged as 2.7.1

2017-03-19 Jan Kotanski <jankotan@gmail.com>
	* change log name to nexus_configuation_logs
	* tagged as 2.7.0

2017-03-06 Jan Kotanski <jankotan@gmail.com>
	* add fixes for H5PYWriter
	* add tests
	* tagged as 2.6.1

2017-03-02 Jan Kotanski <jankotan@gmail.com>
	* add FileWriter class
	* add support for h5py
	* tagged as 2.6.0

2016-11-25 Jan Kotanski <jankotan@gmail.com>
	* add link target as link content
	* tagged as 2.5.0

2016-10-31 Jan Kotanski <jankotan@gmail.com>
	* add nexus_strategy attribute
	* update man page
	* tagged as 2.4.0

2016-09-16 Jan Kotanski <jankotan@gmail.com>
	* add __nxroot__ object into PYEVAL "commonblock"
	* tagged as 2.3.0

2016-09-06 Jan Kotanski <jankotan@gmail.com>
	* add set_source to device proxies
	* tagged as 2.2.6


nxswriter/NXSFromXML.py
#!/usr/bin/env python
#   This file is part of nexdatas - Tango Server for NeXus data writer
#
#    Copyright (C) 2012-2018 DESY, Jan Kotanski <jkotan@mail.desy.de>
#
#    nexdatas is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    nexdatas is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with nexdatas.  If not, see <http://www.gnu.org/licenses/>.
#

""" Command-line tool to ascess to Tango Data Server"""

import sys
import json
import time
import pytz
import datetime
import argparse

from . import TangoDataWriter


class CreateFile(object):

    """ Create File runner"""

    def __init__(self, options):
        """ the main program function

        :param options: parser options
        :type options: :class:`argparse.Namespace`
        """
        #: (:obj:`str`) file name with optional nexus path
        self.__parent = options.parent or ""
        #: (:obj:`str`) xml configuration string
        self.__xml = ""
        #: (:obj:`str`) json data string
        self.__data = options.data
        #: (:obj:`float`) time between records in seconds
        self.__stime = float(options.stime)
        #: (:obj:`int`) number of record steps
        self.__nrecords = int(options.nrecords)
        #: (:obj:`str`) json file
        self.__jsonfile = options.jsonfile
        #: (:obj:`bool`) verbose mode
        self.__verbose = options.verbose or False
        #: (:obj:`bool`) append mode
        self.__append = options.append or False

        if options.args and len(options.args):
            self.__xml = options.args[0].strip()
        else:
            self.__xml = open(options.xmlfile, 'r').read()

        #: (:obj:`str`) writer module
        self.__writer = ""
        if options.h5cpp:
            self.__writer = "h5cpp"
        elif options.h5py:
            self.__writer = "h5py"
        elif options.pni:
            self.__writer = "pni"
        elif "h5cpp" in TangoDataWriter.WRITERS.keys():
            self.__writer = "h5cpp"
        elif "h5py" in TangoDataWriter.WRITERS.keys():
            self.__writer = "h5py"
        else:
            self.__writer = "pni"
        if self.__writer not in TangoDataWriter.WRITERS.keys():
            raise Exception("nxsfromxml: Writer '%s' cannot be found\n"
                            % self.__writer)

    @classmethod
    def currenttime(cls):
        """ returns current time string

        :returns: current time
        :rtype: :obj:`str`
        """
        tzone = time.tzname[0]
        tz = pytz.timezone(tzone)
        fmt = '%Y-%m-%dT%H:%M:%S.%f%z'
        starttime = tz.localize(datetime.datetime.now())
        return str(starttime.strftime(fmt))

    def jsonstring(self):
        """ merges data in json string

        :returns: json string
        :rtype: :obj:`str`
        """
        jsn = {}
        if self.__jsonfile:
            sjsn = open(self.__jsonfile, 'r').read()
            if sjsn.strip():
                jsn = json.loads(sjsn.strip())
        if "data" not in jsn.keys():
            jsn["data"] = {}
        if "start_time" not in jsn["data"]:
            jsn["data"]["start_time"] = self.currenttime()
        if "end_time" not in jsn["data"]:
            jsn["data"]["end_time"] = self.currenttime()
        if str(self.__data.strip()):
            data = json.loads(str(self.__data.strip()))
            jsn["data"].update(data)
        return json.dumps(jsn)

    def run(self):
        """ the main program function
        """
        tdw = TangoDataWriter.TangoDataWriter()
        if self.__verbose:
            print("using the '%s' writer module" % self.__writer)

        tdw.writer = self.__writer
        tdw.fileName = str(self.__parent)
        if self.__verbose:
            print("opening the '%s' file" % self.__parent)
        tdw.openFile()
        tdw.xmlsettings = self.__xml

        if self.__verbose:
            print("opening the data entry")
        tdw.jsonrecord = self.jsonstring()
        tdw.skipacquisition = self.__append
        tdw.openEntry()
        for i in range(self.__nrecords):
            if self.__verbose:
                print("recording step of the H5 file")
            tdw.jsonrecord = self.jsonstring()
            tdw.record()
            if self.__nrecords > 1:
                if self.__verbose:
                    print("sleeping for 1s")
                time.sleep(self.__stime)
        if self.__verbose:
            print("closing the data entry ")
        tdw.jsonrecord = self.jsonstring()
        tdw.skipacquisition = self.__append
        tdw.closeEntry()

        if self.__verbose:
            print("closing the H5 file")
        tdw.closeFile()


def main():
    """ the main program function
    """

    #: pipe arguments
    pipe = ""
    if not sys.stdin.isatty():
        pp = sys.stdin.readlines()
        #: system pipe
        pipe = "".join(pp)

    description = "Command-line tool for creating NeXus files" \
                  + " from XML template"

    epilog = 'examples:\n' \
        '  nxsfromxml  -x config.xml -p newfile.nxs -j datastring.json\n\n' \
        '    - creates a newfile.nxs file with configuration from config.xml' \
        ' and data from datastring.json\n\n' \
        '  nxsfromxml  -x config.xml -p outputfile.nxs -d'\
        ' \'{"sample_name":"sn", "chemical_formula":"H2O", ' \
        '"beamtime_id":"1234", "title":"Sample"}\'\n\n' \
        '    - creates an outputfile.nxs file with configuration ' \
        'from config.xml and data from given json dictionary\n\n' \
        '  nxsconfig get source mono slit1 mirror2 | ' \
        'nxsfromxml -p basicsetup.nxs\n\n' \
        '    - creates an basicsetup.nxs file with configuration ' \
        'of source mono slit1 mirror2 components ' \
        'created by nxsconfig get \n\n'
    parser = argparse.ArgumentParser(
        description=description, epilog=epilog,
        formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument(
        'args', metavar='xml_config', type=str, nargs='?',
        help='nexus writer configuration string')
    parser.add_argument(
        "-x", "--xml-file",
        help="optional file with nexus "
        "configuration string",
        dest="xmlfile", default="")
    parser.add_argument(
        "-p", "--parent",
        help="nexus file name",
        dest="parent", default="")
    parser.add_argument(
        "-d", "--data",
        help="json string with data",
        dest="data", default="")
    parser.add_argument(
        "-j", "--json-file",
        help="json data file in a jsonrecord attribute format",
        dest="jsonfile", default="")
    parser.add_argument(
        "-t", "--time",
        help="time between record steps in seconds, default: 1 s",
        dest="stime", default="1")
    parser.add_argument(
        "-n", "--nrecords",
        help="number of performed record steps",
        dest="nrecords", default="1")
    parser.add_argument(
        "-v", "--verbose", action="store_true",
        default=False, dest="verbose",
        help="verbose mode")
    parser.add_argument(
        "-a", "--append", action="store_true",
        default=False, dest="append",
        help="append mode")
    parser.add_argument(
        "--h5cpp", action="store_true",
        default=False, dest="h5cpp",
        help="use h5cpp module as a nexus reader")
    parser.add_argument(
        "--h5py", action="store_true",
        default=False, dest="h5py",
        help="use h5py module as a nexus reader")
    parser.add_argument(
        "--pni", action="store_true",
        default=False, dest="pni",
        help="use pni module as a nexus reader")

    try:
        options = parser.parse_args()
    except Exception as e:
        sys.stderr.write("Error: %s\n" % str(e))
        sys.stderr.flush()
        parser.print_help()
        print("")
        sys.exit(255)

    #: command-line and pipe arguments
    parg = []
    if hasattr(options, "args"):
        parg = [options.args] if options.args else []
    if pipe:
        parg.append(pipe)
    options.args = parg

    if not options.xmlfile and (
            not options.args or len(options.args) < 1):
        parser.print_help()
        sys.exit(255)
    if options.xmlfile and (
            options.args and len(options.args) > 0):
        parser.print_help()
        sys.exit(255)

    if not options.parent:
        parser.print_help()
        sys.exit(255)

    try:
        command = CreateFile(options)
        command.run()
    except Exception as e:
        sys.stderr.write("Error: nxsfromxml interrupted with:")
        sys.stderr.write(str(e))
        sys.stderr.flush()
        sys.exit(255)


if __name__ == "__main__":
    main()
